crypto/cipher_extra/asm/chacha20_poly1305_x86_64.pl

```diff
@@ -0,0 +1,2468 @@
+        $code=$code.shift(@loop_body)."\n";
+    };
+}
+
+{
+################################################################################
+# void poly_hash_ad_internal();
+$code.="
+.type poly_hash_ad_internal,\@abi-omnipotent
+.align 64
+poly_hash_ad_internal:
+.cfi_startproc
+.cfi_def_cfa rsp, 8
+    xor $acc0, $acc0
+    xor $acc1, $acc1
+    xor $acc2, $acc2
+    cmp \$13,  $itr2
+    jne .Lhash_ad_loop
+.Lpoly_fast_tls_ad:
+    # Special treatment for the TLS case of 13 bytes
+    mov ($adp), $acc0
+    mov 5($adp), $acc1
+    shr \$24, $acc1
+    mov \$1, $acc2\n";
+    &poly_mul(); $code.="
+    ret
+.Lhash_ad_loop:
+        # Hash in 16 byte chunk
+        cmp \$16, $itr2
+        jb .Lhash_ad_tail\n";
+        &poly_add("0($adp)");
+        &poly_mul(); $code.="
+        lea 1*16($adp), $adp
+        sub \$16, $itr2
+    jmp .Lhash_ad_loop
+.Lhash_ad_tail:
+    cmp \$0, $itr2
+    je .Lhash_ad_done
+    # Hash last < 16 byte tail
+    xor $t0, $t0
+    xor $t1, $t1
+    xor $t2, $t2
+    add $itr2, $adp
+.Lhash_ad_tail_loop:
+        shld \$8, $t0, $t1
+        shl \$8, $t0
+        movzxb -1($adp), $t2
+        xor $t2, $t0
+        dec $adp
+        dec $itr2
+    jne .Lhash_ad_tail_loop
+
+    add $t0, $acc0
+    adc $t1, $acc1
+    adc \$1, $acc2\n";
+    &poly_mul(); $code.="
+    # Finished AD
+.Lhash_ad_done:
+    ret
+.cfi_endproc
+.size poly_hash_ad_internal, .-poly_hash_ad_internal\n";
+}
+
+{
+################################################################################
+# void GFp_chacha20_poly1305_open(uint8_t *in_out, size_t len_in, uint8_t *ad, size_t len_ad, uint8_t keyp[48], size_t out_offset);
+$code.="
+.globl GFp_chacha20_poly1305_open
+.type GFp_chacha20_poly1305_open,\@function,6
+.align 64
+GFp_chacha20_poly1305_open:
+.cfi_startproc
+    push %rbp
+.cfi_adjust_cfa_offset 8
+    push %rbx
+.cfi_adjust_cfa_offset 8
+    push %r12
+.cfi_adjust_cfa_offset 8
+    push %r13
+.cfi_adjust_cfa_offset 8
+    push %r14
+.cfi_adjust_cfa_offset 8
+    push %r15
+.cfi_adjust_cfa_offset 8
+    push $keyp
+.cfi_adjust_cfa_offset 8
+    sub \$288 + $xmm_storage + 32, %rsp
+.cfi_adjust_cfa_offset 288 + 32
+.cfi_offset rbp, -16
+.cfi_offset rbx, -24
+.cfi_offset r12, -32
+.cfi_offset r13, -40
+.cfi_offset r14, -48
+.cfi_offset r15, -56
+    lea 32(%rsp), %rbp
+    and \$-32, %rbp\n";
+$code.="
+    movaps %xmm6,16*0+$xmm_store
+    movaps %xmm7,16*1+$xmm_store
+    movaps %xmm8,16*2+$xmm_store
+    movaps %xmm9,16*3+$xmm_store
+    movaps %xmm10,16*4+$xmm_store
+    movaps %xmm11,16*5+$xmm_store
+    movaps %xmm12,16*6+$xmm_store
+    movaps %xmm13,16*7+$xmm_store
+    movaps %xmm14,16*8+$xmm_store
+    movaps %xmm15,16*9+$xmm_store\n" if ($win64);
+$code.="
+    mov $adl, 0+$len_store
+    mov $inl, 8+$len_store\n";
+$code.="
+    mov GFp_ia32cap_P+8(%rip), %eax
+    and \$`(1<<5) + (1<<8)`, %eax # Check both BMI2 and AVX2 are present
+    xor \$`(1<<5) + (1<<8)`, %eax
+    jz chacha20_poly1305_open_avx2\n" if ($avx>1);
```

Comment:
In the calling Rust code, we already do feature detection to get this far. It seems like, especially since we're already changing the API of the assembly code, we should change the assembly API so that the Rust code calls `chacha20_poly1305_open_sse41_avx2` if BMI2 and AVX2 are present, and otherwise calls `GFp_chacha20_poly1305_open`, which should be renamed to `GFp_chacha20_poly1305_open_sse41`.

Also, how important it is to retain the non-AVX2+BMI2 SSE4.1 implementation in the first place? Why not just have the Rust code check for AVX2+BMI2+SSE4.1 and then call chacha20_poly1305_open_avx2, or else fall back to the previously-existing code? Are machines that have SSE4.1 but don't have AVX2 important for what you're doing?

Reply:
> we should change the assembly API so that the Rust code calls `chacha20_poly1305_open_sse41_avx2` if BMI2 and AVX2 are present

That is doable, but will result in some code duplication. Also the avx2 function returns to the use function to handle the remainder, which I would also have to duplicate. In addition checking `GFp_ia32cap_P` is performed in all of the current asm files: `aesni-x86_64.pl`, `x86_64-mont5.pl`, `chacha-x86_64.pl`, `p256-x86_64-asm.pl` and others.

>  how important it is to retain the non-AVX2+BMI2 SSE4.1 implementation in the first place

To my knowledge Intel still produces processors without AVX2 support. I would prefer to support low end devices as well. I don't have access to one right now to measure the performance delta, but I suspect it is even more significant than for AVX2 enabled devices.

Reply:
Thanks. I think it makes sense to leave it as-is for now.